src/core/random.hpp

```diff
@@ -148,30 +141,45 @@ Utils::Vector3d v_noise(uint64_t counter, int key1, int key2 = 0) {
  * @param key2 key for random number generation
  * @tparam salt (decorrelates different thermostat types)
  *
- * @return 3D vector of Gaussian random numbers.
+ * @return Vector of Gaussian random numbers.
  *
  */
-template <RNGSalt salt>
-inline Utils::Vector3d v_noise_g(uint64_t counter, int key1, int key2 = 0) {
-
-  auto const noise = philox_4_uint64s<salt>(counter, key1, key2);
-  using Utils::uniform;
-  Utils::Vector4d u{uniform(noise[0]), uniform(noise[1]), uniform(noise[2]),
-                    uniform(noise[3])};
+template <RNGSalt salt, size_t N = 3,
+          typename std::enable_if<(N >= 1) and (N <= 4), int>::type = 0>
+auto noise_gaussian(uint64_t counter, int key1, int key2 = 0) {
 
-  // Replace zeros from uniformly distributed numbers (see doc string)
+  auto const integers = philox_4_uint64s<salt>(counter, key1, key2);
   static const double epsilon = std::numeric_limits<double>::min();
-  for (int i = 0; i < 4; i++)
-    if (u[i] < epsilon)
-      u[i] = epsilon;
 
-  // Box muller transform code adapted from
-  // https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform
+  constexpr size_t M = (N <= 2) ? 2 : 4;
+  Utils::VectorXd<M> u{};
+  std::transform(integers.begin(), integers.begin() + M, u.begin(),
+                 [](size_t value) {
+                   auto u = Utils::uniform(value);
+                   return (u < epsilon) ? epsilon : u;
+                 });
 
-  static const double two_pi = 2.0 * Utils::pi();
-  return {sqrt(-2.0 * log(u[0])) * cos(two_pi * u[1]),
-          sqrt(-2.0 * log(u[0])) * sin(two_pi * u[1]),
-          sqrt(-2.0 * log(u[2])) * cos(two_pi * u[3])};
+  // Box-Muller transform code adapted from
+  // https://en.wikipedia.org/wiki/Box%E2%80%93Muller_transform
+  // optimizations: the modulo is cached (logarithms are expensive), the
+  // sin/cos are evaluated simultaneously by gcc or separately by Clang
+  Utils::VectorXd<N> noise{};
+  constexpr double two_pi = 2.0 * Utils::pi();
+  auto modulo = sqrt(-2.0 * log(u[0]));
```

Comment:
Please don't reuse variables like this. Make `modulo` and and `angle` const, and introduce new variables in the inner scope. You can get this by nesting the if's like

```c++
if(N > 2) {
  auto const module = ...
 ...
 if(N > 3) {
 ....
}
}
```
because `(N > 2)` is a special case of `(N > 3)`.